Tests/UnitTests/Diagnostics/DiagnosticsSynchronizerTests.swift

```diff
@@ -0,0 +1,174 @@
+    }
+
+    func testSyncOneEvent() async throws {
+        let event = await self.storeEvent()
+
+        try await self.synchronizer.syncDiagnosticsIfNeeded()
+
+        expect(self.api.invokedPostDiagnosticsEvents) == true
+        expect(self.api.invokedPostDiagnosticsEventsParameters) == [[
+            event
+        ]]
+
+        await self.verifyEmptyStore()
+    }
+
+    func testSyncMultipleEvents() async throws {
+        let event1 = await self.storeEvent()
+        let event2 = await self.storeEvent(timestamp: Self.eventTimestamp2)
+
+        try await self.synchronizer.syncDiagnosticsIfNeeded()
+
+        expect(self.api.invokedPostDiagnosticsEvents) == true
+        expect(self.api.invokedPostDiagnosticsEventsParameters) == [[ event1, event2 ]]
+
+        await self.verifyEmptyStore()
+    }
+
+    func testSyncWithUnsuccessfulError() async throws {
+        let event = await self.storeEvent()
+        let expectedError: NetworkError = .offlineConnection()
+
+        self.api.stubbedPostDiagnosticsEventsCompletionResult = .networkError(expectedError)
+        do {
+            _ = try await self.synchronizer.syncDiagnosticsIfNeeded()
+            fail("Expected error")
+        } catch BackendError.networkError(expectedError) {
+            // Expected
+        } catch {
+            throw error
+        }
+
+        expect(self.api.invokedPostDiagnosticsEvents) == true
+        expect(self.api.invokedPostDiagnosticsEventsParameters) == [[ event ]]
+
+        await self.verifyEvents([event])
+    }
+
+    func testSyncWithSuccessfullySyncedError() async throws {
+        _ = await self.storeEvent()
+        _ = await self.storeEvent(timestamp: Self.eventTimestamp2)
+
+        let expectedError: NetworkError = .errorResponse(.defaultResponse, .invalidRequest)
+
+        self.api.stubbedPostDiagnosticsEventsCompletionResult = .networkError(expectedError)
+        do {
+            _ = try await self.synchronizer.syncDiagnosticsIfNeeded()
+            fail("Expected error")
+        } catch BackendError.networkError(.errorResponse) {
+            // Expected
+        } catch {
+            throw error
+        }
+
+        expect(self.api.invokedPostDiagnosticsEvents) == true
+
+        await self.verifyEmptyStore()
+    }
+
+    func testCannotSyncMultipleTimesInParallel() async throws {
+        // The way this test is written does not work in iOS 15.
+        // The second Task does not start until the first one is done.
+        try AvailabilityChecks.iOS16APIAvailableOrSkipTest()
+
+        _ = await self.storeEvent()
+        _ = await self.storeEvent(timestamp: Self.eventTimestamp2)
+
+        let task1 = Task { [synchronizer = self.synchronizer!] in try await synchronizer.syncDiagnosticsIfNeeded() }
+        let task2 = Task { [synchronizer = self.synchronizer!] in try await synchronizer.syncDiagnosticsIfNeeded() }
+
+        let (result1: (), result2: ()) = try await (task1.value, task2.value)
```

Comment:
Maybe slightly simpler to read?

```suggestion
        let syncer = self.synchronizer!
        async let sync1: () = await syncer.syncDiagnosticsIfNeeded()
        async let sync2: () = await syncer.syncDiagnosticsIfNeeded()
        
        let (result1: (), result2: ()) = try await (sync1, sync2)
```

Reply:
Thanks for the suggestion!